platform/core/src/services/StudyPrefetcherService/StudyPrefetcherService.ts

```diff
@@ -0,0 +1,582 @@
+
+const IMAGE_REQUEST_TYPE = 'prefetch';
+
+export const EVENTS = {
+  SERVICE_STARTED: 'event::studyPrefetcherService:started',
+  SERVICE_STOPPED: 'event::studyPrefetcherService:stopped',
+  DISPLAYSET_LOAD_PROGRESS: 'event::studyPrefetcherService:displaySetLoadProgress',
+  DISPLAYSET_LOAD_COMPLETE: 'event::studyPrefetcherService:displaySetLoadComplete',
+};
+
+enum StudyPrefetchOrder {
+  closest = 'closest',
+  topBottom = 'top-bottom',
+  bottomUp = 'bottom-up',
+}
+
+type StudyPrefetchConfig = {
+  enabled: boolean;
+  displaySetCount: number;
+  maxNumPrefetchRequests: number;
+  order: StudyPrefetchOrder;
+};
+
+type DisplaySetStatus = {
+  displaySetInstanceUID: string;
+  numImages: number;
+  // numImagesLoaded: number;
+
+  pendingImageIds: Set<string>;
+  loadedImageIds: Set<string>;
+  failedImageIds: Set<string>;
+};
+
+type ImageRequest = {
+  displaySetInstanceUID: string;
+  imageId: string;
+  aborted: boolean;
+};
+
+type PubSubServiceSubscription = { unsubscribe: () => any };
+
+class StudyPrefetcherService extends PubSubService {
+  private _extensionManager: ExtensionManager;
+  private _servicesManager: ServicesManager;
+  private _subscriptions: PubSubServiceSubscription[];
+  private _activeDisplaySetsInstanceUIDs: string[] = [];
+  // private _studies: any[];
+  // private _displaySets: any[];
+  // private _sortedDisplaySets: DisplaySet[];
+  private _pendingRequests: ImageRequest[] = [];
+  private _inflightRequests = new Map<string, ImageRequest>();
+  // private _nextDisplaySetLoadingIndex: number;
+  private _isRunning = false;
+  // private _hasDisplaySetLoadingInProgress = false;
+  // private _loadedDisplaySets = new Set();
+  private _displaySetLoadingStates = new Map<string, DisplaySetStatus>();
+  private config: StudyPrefetchConfig = {
+    enabled: false,
+    displaySetCount: 3,
+    maxNumPrefetchRequests: 1000,
+    order: StudyPrefetchOrder.bottomUp,
+  };
+
+  public static REGISTRATION = {
+    name: 'studyPrefetcherService',
+    altName: 'StudyPrefetcherService',
+    create: ({ configuration, servicesManager, extensionManager }): StudyPrefetcherService => {
+      return new StudyPrefetcherService({
+        servicesManager,
+        extensionManager,
+        configuration,
+      });
+    },
+  };
+
+  constructor({
+    servicesManager,
+    extensionManager,
+    configuration,
+  }: {
+    servicesManager: ServicesManager;
+    extensionManager: ExtensionManager;
+    configuration: StudyPrefetchConfig;
+  }) {
+    super(EVENTS);
+
+    this._servicesManager = servicesManager;
+    this._extensionManager = extensionManager;
+    this._subscriptions = [];
+
+    Object.assign(this.config, configuration);
+    // this._displaySets = [];
+
+    // DEBUG - DELETE
+    setTimeout(() => this.start(), 1000);
+  }
+
+  // Move this method to a different place (component)
+  private start() {
+    const { displaySetService, viewportGridService } = this._servicesManager.services;
```

Comment:
@sedghi  Where would be a better place to add some listeners? (UI <-- listeners --> StudyPrefetcher)